### PR TITLE
Remove support for UTF-7

### LINKED ISSUE #8363: [NBug] Input string was not in a correct format.
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.FormatException: Input string was not in a correct format.
   at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   at GitUI.Editor.Diff.DiffLineNumAnalyzer.Analyze(String diffContent)
   at GitUI.Editor.Diff.DiffViewerLineNumberControl.DisplayLineNumFor(String diff)
   at GitUI.Editor.FileViewerInternal.SetText(String text, Action openWithDifftool, Boolean isDiff)
   at GitUI.Editor.FileViewer.<>c__DisplayClass103_0.<ViewPatchAsync>b__0()
   at GitUI.Editor.FileViewer.ShowOrDeferAsync(Int64 contentLength, Func`1 showFunc)
   at GitUI.Editor.FileViewer.<ViewPatchAsync>d__103.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.CommandsDialogs.RevisionDiffControl.<ShowSelectedFileDiffAsync>d__35.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.CommandsDialogs.RevisionDiffControl.<<DiffText_ExtraDiffArgumentsChanged>b__39_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
```

## Additional information
Not sure what I did...nothing special...
But I updated git to version 2.28.0 earlier today...

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.26.2.windows.1
- Microsoft Windows NT 10.0.18363.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

### LINKED ISSUE #8367: [NBug] Input string was not in a correct format.
I had a quick look for existing issues, and while a couple have matching exceptions, the steps to reproduce seem to be very different.

## Current behaviour
Selecting Unicode (UTF-7) as the encoding for the diff viewer results in a crash

## Expected behaviour
No crash

## Steps to reproduce
1. Open repo
1. Select a commit to view details in Commit tab
1. Swap to Diff tab
1. Mouse over diff window to reveal diff viewer controls
1. Open text encoding dropdown (currently selected: Unicode (UTF-8)
1. Select Unicode (UTF-7)

image

## Error Details
```
System.FormatException: Input string was not in a correct format.
   at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   at GitUI.Editor.Diff.DiffLineNumAnalyzer.Analyze(String diffContent)
   at GitUI.Editor.Diff.DiffViewerLineNumberControl.DisplayLineNumFor(String diff)
   at GitUI.Editor.FileViewerInternal.SetText(String text, Action openWithDifftool, Boolean isDiff)
   at GitUI.Editor.FileViewer.<>c__DisplayClass103_0.<ViewPatchAsync>b__0()
   at GitUI.Editor.FileViewer.ShowOrDeferAsync(Int64 contentLength, Func`1 showFunc)
   at GitUI.Editor.FileViewer.<ViewPatchAsync>d__103.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.CommandsDialogs.RevisionDiffControl.<ShowSelectedFileDiffAsync>d__35.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.CommandsDialogs.RevisionDiffControl.<<DiffText_ExtraDiffArgumentsChanged>b__39_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
```

## Additional information
In the Diff changes window, opened the encoding dropdown and changed from Unicode (UTF-8) to Unicode (UTF-7)

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.23.0.windows.1 (recommended: 2.25.1 or later)
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4180.0
- DPI 96dpi (no scaling)

### LINKED ISSUE #8425: [NBug] Input string was not in a correct format.
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.FormatException: Input string was not in a correct format.
   at System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   at System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   at GitUI.Editor.Diff.DiffLineNumAnalyzer.Analyze(String diffContent)
   at GitUI.Editor.Diff.DiffViewerLineNumberControl.DisplayLineNumFor(String diff)
   at GitUI.Editor.FileViewerInternal.SetText(String text, Action openWithDifftool, Boolean isDiff)
   at GitUI.Editor.FileViewer.<>c__DisplayClass103_0.<ViewPatchAsync>b__0()
   at GitUI.Editor.FileViewer.ShowOrDeferAsync(Int64 contentLength, Func`1 showFunc)
   at GitUI.Editor.FileViewer.<ViewPatchAsync>d__103.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.UserControls.CommitDiff.<ViewSelectedDiffAsync>d__9.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.UserControls.CommitDiff.<<DiffText_ExtraDiffArgumentsChanged>b__8_0>d.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
```

## Additional information
I just used mouse scroll button and this error window popped up..

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.23.0.windows.1 (recommended: 2.25.1 or later)
- Microsoft Windows NT 10.0.18362.0
- .NET Framework 4.8.4200.0
- DPI 96dpi (no scaling)

### LINKED ISSUE #8634: [NBug] Die Eingabezeichenfolge hat das falsche Format.
:warning: The sections below must be filled in and this text must be removed or the issue will be closed.

## Error Details
```
System.FormatException: Die Eingabezeichenfolge hat das falsche Format.
   bei System.Number.StringToNumber(String str, NumberStyles options, NumberBuffer& number, NumberFormatInfo info, Boolean parseDecimal)
   bei System.Number.ParseInt32(String s, NumberStyles style, NumberFormatInfo info)
   bei GitUI.Editor.Diff.DiffLineNumAnalyzer.Analyze(String diffContent)
   bei GitUI.Editor.Diff.DiffViewerLineNumberControl.DisplayLineNumFor(String diff)
   bei GitUI.Editor.FileViewerInternal.SetText(String text, Action openWithDifftool, Boolean isDiff)
   bei GitUI.Editor.FileViewer.<>c__DisplayClass103_0.<ViewPatchAsync>b__0()
   bei GitUI.Editor.FileViewer.ShowOrDeferAsync(Int64 contentLength, Func`1 showFunc)
   bei GitUI.Editor.FileViewer.<ViewPatchAsync>d__103.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei GitUI.CommandsDialogs.RevisionDiffControl.<ShowSelectedFileDiffAsync>d__35.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei GitUI.CommandsDialogs.RevisionDiffControl.<<DiffText_ExtraDiffArgumentsChanged>b__39_0>d.MoveNext()
--- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
   bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   bei GitUI.ThreadHelper.<>c__DisplayClass13_0.<<FileAndForget>b__0>d.MoveNext()
```

## Additional information
Selected a commit and went to "Diff". Upper right corner of code window, changed code formating from UTF-8 to "Unicode (UTF-7)".

## Environment
- Git Extensions 3.4.3.9999
- Build d4b0f48bbf3e39a71f5d7ff5231be5678d4aa0f1
- Git 2.28.0.windows.1
- Microsoft Windows NT 10.0.19041.0
- .NET Framework 4.8.4250.0
- DPI 96dpi (no scaling)

### PR DESCRIPTION
**Pull request by RussKie**
See: https://github.com/dotnet/docs/issues/19274

Fixes #8363
Fixes #8367
Fixes #8425
Fixes #8634

### Before

image
image

### After

image
image

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

File: GitCommands/Settings/AppSettings.cs [hunk 1]
```diff
@@ -1899,6 +1903,12 @@ void AddEncodingByName(string s)
                 var utf8 = new UTF8Encoding(false);
                 foreach (var encodingName in availableEncodings.Split(';'))
                 {
+                    if (encodingName == "utf-7")
```
Comment (ivangrek) [inline GitCommands/Settings/AppSettings.cs:1906]: Encoding.UTF7

File: GitUI/CommandsDialogs/SettingsDialog/FormAvailableEncodings.cs [hunk 2]
```diff
@@ -30,6 +30,7 @@ private void LoadEncoding()
             }
 
             var availableEncoding = Encoding.GetEncodings()
+                .Where(ei => ei.Name != "utf-7") // UTF-7 is no longer supported, see: https://github.com/dotnet/docs/issues/19274
```
Comment (ivangrek) [inline GitUI/CommandsDialogs/SettingsDialog/FormAvailableEncodings.cs:33]: Encoding.UTF7

Comment (RussKie) [reply GitUI/CommandsDialogs/SettingsDialog/FormAvailableEncodings.cs]: Thanks

File: GitCommands/Settings/AppSettings.cs (hunk 1 above)
Comment (RussKie) [reply GitCommands/Settings/AppSettings.cs]: Thanks

**Review by ivangrek — APPROVED** [review]
👍

**Comment by RussKie** [comment]
@msftbot merge in 27 hours

**Pull request merged**